### PR TITLE
Add `save_on_exception` option to `ModelCheckpoint`

### LINKED ISSUE #19686: `dirpath` in `OnExceptionCheckpoint` should default to the same directory as `ModelCheckpoint`
### Description & Motivation

`ModelCheckpoint` is a very handy features and `OnExceptionCheckpoint` is nice to have when you want to interrupt a training or there is a bug. However there is no is no default value for `dirpath` in `OnExceptionCheckpoint`, it would be nice if it defaulted to the same directory as `ModelCheckpoint` which is evaluated at runtime and under the logging directory.

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by vsey**
What does this PR do?
This PR adds a `save_on_exception` option to the ModelCheckpoint callback. This some of this functionality is already implemented in the `OnExceptionCheckpoint` checkpoint, but I believe that bundling all checkpoint options in the ModelCheckpoint is more intuitive. Additionally, this leads to the same naming conventions and directory paths used for the exception checkpoint as for all the others.
When enabled, this option serves as a contingency in case of any disruption during training, allowing one to continue from the last step before the exception occurs without losing too much progress. By printing the exception type and message, this also alleviates issue #20187.

Fixes #19686 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20916.org.readthedocs.build/en/20916/

### CODE REVIEW

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -338,6 +342,19 @@ def on_validation_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModul
...
+        """Save a checkpoint when an exception is raised."""
+        if self._should_save_on_exception(trainer):
+            monitor_candidates = self._monitor_candidates(trainer)
+            filepath = self.format_checkpoint_name(metrics=monitor_candidates, prefix=self.CHECKPOINT_EXCEPTION_PREFIX)
+            self._save_checkpoint(trainer, filepath)
+            self._save_last_checkpoint(trainer, monitor_candidates)
+            rank_zero_info(
+                f"An {type(exception).__name__} was raised with message: \
+                {str(exception)}, saved checkpoint to {filepath}"
+            )
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/model_checkpoint.py]:
```suggestion
        if not self._should_save_on_exception(trainer):
            return
        monitor_candidates = self._monitor_candidates(trainer)
        filepath = self.format_checkpoint_name(metrics=monitor_candidates, prefix=self.CHECKPOINT_EXCEPTION_PREFIX)
        self._save_checkpoint(trainer, filepath)
        self._save_last_checkpoint(trainer, monitor_candidates)
        rank_zero_info(
            f"An {type(exception).__name__} was raised with message: \
            {str(exception)}, saved checkpoint to {filepath}"
        )
```

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -764,6 +770,306 @@ def test_ckpt_every_n_train_steps(tmp_path):
...
+    class TroubledModelOnValidationEnd(BoringModel):
+        def on_validation_end(self):
+            if not self.trainer.sanity_checking:
+                raise RuntimeError("Trouble!")
+
+    class TroubledModelOnFitEnd(BoringModel):
+        def on_fit_end(self):
+            raise RuntimeError("Trouble!")
+
+    models = [
```
Comment (Borda) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py]: lets have this as test parametrisation

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -213,6 +214,7 @@ class ModelCheckpoint(Checkpoint):
     CHECKPOINT_JOIN_CHAR = "-"
     CHECKPOINT_EQUALS_CHAR = "="
     CHECKPOINT_NAME_LAST = "last"
+    CHECKPOINT_EXCEPTION_PREFIX = "exception"
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/model_checkpoint.py]: Thinking g we need a new prefix since in fact it will be the last

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py (hunk 2 above)
Comment (vsey) [reply tests/tests_pytorch/checkpointing/test_model_checkpoint.py]: Should I move the models and callback for testing in a separate file? Because in my opinion they clutter the test file a bit too much

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -578,13 +602,17 @@ def _format_checkpoint_name(
...
 
         return filename
 
     def format_checkpoint_name(
-        self, metrics: dict[str, Tensor], filename: Optional[str] = None, ver: Optional[int] = None
+        self,
+        metrics: dict[str, Tensor],
+        filename: Optional[str] = None,
+        prefix: Optional[str] = None,
+        ver: Optional[int] = None,
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/model_checkpoint.py]:
```suggestion
        ver: Optional[int] = None,
        prefix: Optional[str] = None,
```
to prevent breaking chnages

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -616,7 +644,9 @@ def format_checkpoint_name(
 
         """
         filename = filename or self.filename
-        filename = self._format_checkpoint_name(filename, metrics, auto_insert_metric_name=self.auto_insert_metric_name)
+        filename = self._format_checkpoint_name(
+            filename, metrics, prefix, auto_insert_metric_name=self.auto_insert_metric_name
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/model_checkpoint.py]:
```suggestion
            filename, metrics, prefix=prefix, auto_insert_metric_name=self.auto_insert_metric_name
```

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 6]
```diff
@@ -764,6 +770,420 @@ def test_ckpt_every_n_train_steps(tmp_path):
...
+    )
+    trainer = Trainer(
+        default_root_dir=tmp_path, limit_train_batches=2, callbacks=[checkpoint_callback], max_epochs=2, logger=False
+    )
+
+    with pytest.raises(RuntimeError, match="Trouble!"):
+        trainer.fit(model)
+
+    assert not os.path.isfile(tmp_path / "exception-already_saved.ckpt")
+    assert os.path.isfile(tmp_path / "already_saved.ckpt")
```
Comment (Borda) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py]:
```suggestion
```
This seems to be the same as the parametrization bellow

Comment (vsey) [reply tests/tests_pytorch/checkpointing/test_model_checkpoint.py]: Thank you for your feedback. I just wanted to clarify the differences between the two tests.

The first test I use to check the run conditions and to ensure that no checkpoint gets created during sanity checks , fast dev runs or if a checkpoint was already saved at the current training step.

While the other test check that if the run conditions are true (i.e. no sanity check or fast dev run) and an exception is raised in any section of the training process, the checkpoint gets saved correctly

Also, I change the description for the first test, to better clarify this, as it was very cryptic before

Comment (Borda) [reply tests/tests_pytorch/checkpointing/test_model_checkpoint.py]: ok, so lest parametrize the fist one too :)

**Review by Borda — APPROVED** [review]
